tensorflow/c/experimental/pluggable_profiler/pluggable_profiler.h

```diff
@@ -0,0 +1,169 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#ifndef TENSORFLOW_C_EXPERIMENTAL_PLUGGABLE_PROFILER_PLUGGABLE_PROFILER_H_
+#define TENSORFLOW_C_EXPERIMENTAL_PLUGGABLE_PROFILER_PLUGGABLE_PROFILER_H_
+#include <stddef.h>
+#include <stdint.h>
+
+#include "tensorflow/c/c_api_macros.h"
+#include "tensorflow/c/tf_status.h"
+
+// C API for Pluggable Profiler. The API is under active development and
+// eventually should allow registering a profiler with TensorFlow.
+//
+// Conventions:
+//   * Struct prefix indicates whether struct fields should be filled by the
+//   plug-in or core TensorFlow implementation:
+//     * TF_: Set/filled by core, unless marked otherwise.
+//     * TP_: Set/filled by plug-in, unless marked otherwise.
+//     * This prefix rule only applies to structures. Enumerations and methods
+//     are all prefixed with TP_.
+//   * Structs begin with two fields:
+//     * size_t struct_size: Stores the unpadded size of the struct.
+//     * void* ext: A reserved field that may be populated by a plugin in TP_*
+//     structs or potential future extension points in TF_ structs. Must be set
+//     to zero by default if it unused.
+//  * We use struct_size for version checking by both core and plug-in.
+//    * It is exempt from the TF/TP rule above and must be set both by core and
+//    plug-in.
+//    * It can be checked programmatically to determine which struct fields are
+//    available in the structure.
+//  * When a member is added to a struct, the struct size definition must be
+//  updated to use the new last member of the struct.
+//
+// Example usage:
+//   /* Sample TensorFlow code below, exact implementation might differ. */
+//   // Version checking uses `struct_size`. It is exempt from the `TF/TP` rule
+//   // above and should be set both by core and the plugin."
+//
+//   /* Plugin code below */
+//   void profiler_start(const TP_Profiler* profiler, TF_Status* status) {
+//     /* Enable profiler */
+//     ...
+//   }
+//
+//  void profiler_stop(const TP_Profiler* profiler, TF_Status* status) {
+//    /* Disable Profiler */
+//    ...
+//  }
+//
+//  void profiler_collect_data_xspace(const TP_Profiler* profiler, uint8_t*
+//  buffer, size_t* size_in_bytes, TF_Status* status) {
+//    Xspace xspace = get_my_xspace(); /* Plugin generates Xspace based on
+//    collected profiler data. */ size_t buffer_size_in_bytes = *size_in_bytes;
+//    *size_in_bytes = xspace.ByteSizeLong(); /* get the size of Xspace */
+//    if (buffer == nullptr) {
+//      return; /* TensorFlow will first get the size of Xspace, then allocate
+//      the big enough buffer and pass it to the plugin for retrieving Xspace.
+//      */
+//    }
+//    bool success = xspace.SerializeToArray(buffer, buffer_size_in_bytes);
+//    }
```

Comment:
Minor formatting nit:
```suggestion
//  void profiler_collect_data_xspace(const TP_Profiler* profiler, uint8_t*
//  buffer, size_t* size_in_bytes, TF_Status* status) {
//    /* Plugin generates Xspace based on collected profiler data. */
//    Xspace xspace = get_my_xspace();
//    size_t buffer_size_in_bytes = *size_in_bytes;
//    *size_in_bytes = xspace.ByteSizeLong(); /* get the size of Xspace */
//    if (buffer == nullptr) {
//      /* TensorFlow will first get the size of Xspace, then allocate the big
//         enough buffer and pass it to the plugin for retrieving Xspace. */
//      return;
//    }
//    bool success = xspace.SerializeToArray(buffer, buffer_size_in_bytes);
//  }
```